system/lib/libc/dynlink.c

```diff
@@ -105,7 +107,221 @@ static struct dso* load_library_start(const char* name, int flags) {
+  struct thread_sync_data* data = (struct thread_sync_data*)arg;
+#ifdef DYLINK_DEBUG
+  _emscripten_errf("%p: do_thread_sync: %d", pthread_self(), data->promise_id);
+#endif
+  data->result = _emscripten_thread_sync_code();
+}
+
+// Called once _emscripten_proxy_sync_code completes
+static void done_thread_sync(void* arg) {
+  struct thread_sync_data* data = (struct thread_sync_data*)arg;
+#ifdef DYLINK_DEBUG
+  _emscripten_errf("%p: done_thread_sync: promise_id=%d success=%d",
+                   pthread_self(),
+                   data->promise_id,
+                   data->result);
+#endif
+  if (data->result) {
+    _emscripten_promise_resolve(data->promise_id, NULL);
+  } else {
+    _emscripten_promise_reject(data->promise_id);
+  }
+  free(data);
+}
+
+static em_proxying_queue* dlopen_proxying_queue;
+
+__attribute__((constructor)) static void create_queue() {
+  dlopen_proxying_queue = em_proxying_queue_create();
+}
+
+void _emscripten_process_dlopen_queue() {
+  assert(dlopen_proxying_queue);
+  assert(!emscripten_is_main_runtime_thread());
+  emscripten_proxy_execute_queue(dlopen_proxying_queue);
+}
+
+// Asynchronously runs _emscripten_thread_sync_code on the target then and
+// resolves (or rejects) the given promise once its is complete.
+int _emscripten_proxy_sync_code(pthread_t target_thread, int promise_id) {
+  em_proxying_queue* q = emscripten_proxy_get_system_queue();
+  struct thread_sync_data* data = malloc(sizeof(struct thread_sync_data));
+  data->promise_id = promise_id;
+  data->result = false;
+  return emscripten_proxy_async_with_callback(dlopen_proxying_queue,
+                                              target_thread,
+                                              do_thread_sync,
+                                              data,
+                                              done_thread_sync,
+                                              data);
+}
+
+static void done_sync_all(em_proxying_ctx* ctx) {
+#ifdef DYLINK_DEBUG
+  _emscripten_errf("%p: done_sync_all", pthread_self());
+#endif
+  emscripten_proxy_finish(ctx);
+}
+
+static void main_thread_sync_all(em_proxying_ctx* ctx, void* arg) {
+  pthread_t calling_thread = (pthread_t)arg;
+#ifdef DYLINK_DEBUG
+  _emscripten_errf("%p: main_thread_sync_all calling=%p", pthread_self(), calling_thread);
+#endif
+  _emscripten_sync_all_threads(calling_thread, done_sync_all, ctx);
+}
+
+static void sync_code_all_threads() {
+  // Call `main_thread_sync_all` on the main thread and block until its
+  // complete. This gets called after a shared library is loaded by a worker.
+  pthread_t main_thread = emscripten_main_browser_thread_id();
+#ifdef DYLINK_DEBUG
+  _emscripten_errf("%p: sync_code_all_threads main=%p", pthread_self(), main_thread);
+#endif
+  assert(pthread_self() != main_thread);
+  em_proxying_queue* q = emscripten_proxy_get_system_queue();
+  int success = emscripten_proxy_sync_with_ctx(
+    q, main_thread, main_thread_sync_all, pthread_self());
```

Comment:
Is it possible for two threads to deadlock because they are both stuck in `sync_code_all_threads` waiting for the other to catch up?

Reply:
No, there is an outer lock in `dlopen` which is the only place that `sync_code_all_threads` is called.   Only one thread can enter `dlopen` or call `sync_code_all_threads` at a given time.

Reply:
I guess that works as long as all threads pump the system queue while they wait for locks.

Reply:
Yes, with this change all threads pump the "dlopen" queue while waiting and the main thread pumps the "system" queue. 

IIUC @juj didn't want to system queue to be pumped while block on background worker (like it is for the main thread).     That is one of the nice things about having multiple queues is that i have pump just the dlopen queue and not the rest.

Reply:
This change fundamentally relies on being able to have all thread process the dlopen queue no matter what they are doing.  Obviously if they are spin-locking that won't work, and the thread calling `dlopen` would block until the spin lock was done.